File: python_imp/imp.py

import time
current_milli_time = lambda: int(round(time.time() * 1000))

class KItem(object):
    pass

class ACon(KItem):
    def __init__(self, n):
        self.n = n

class AVar(KItem):
    def __init__(self, v):
        self.v = v

class Div(KItem):
    def __init__(self, x, y):
        self.x = x 
        self.y = y 

class Add(KItem):
    def __init__(self, x, y):
        self.x = x 
        self.y = y

class BCon(KItem):
    def __init__(self, b):
        self.b = b

class Le(KItem):
    def __init__(self, x, y):
        self.x = x 
        self.y = y 

class Not(KItem):
    def __init__(self, b):
        self.b = b 

class And(KItem):
    def __init__(self, x, y):
        self.x = x 
        self.y = y

class Assign(KItem):
    def __init__(self, variable, value):
        self.variable = variable
        self.value = value 

class If(KItem):
    def __init__(self, condition, subsequent, alternative):
        self.condition = condition
        self.subsequent = subsequent 
        self.alternative = alternative

class While(KItem):
    def __init__(self, condition, body):
        self.condition = condition
        self.body = body 

class Seq(KItem):
    def __init__(self, s1, s2):
        self.s1 = s1 
        self.s2 = s2 

class Skip(KItem):
    def __init__(self):
        pass

class Pgm(KItem):
    def __init__(self, vars, body):
        self.vars = vars
        self.body = body

class DivL(KItem):
    def __init__(self, v):
        self.v = v 

class DivR(KItem):
    def __init__(self, v):
        self.v = v 

class AddL(KItem):
    def __init__(self, v):
        self.v = v 

class AddR(KItem):
    def __init__(self, v):
        self.v = v 

class LeL(KItem):
    def __init__(self, v):
        self.v = v 

class LeR(KItem):
    def __init__(self, v):
        self.v = v 

class NotF(KItem):
    def __init__(self):
        pass

class AndL(KItem):
    def __init__(self, v):
        self.v = v 

class AssignR(KItem):
    def __init__(self, variable):
        self.variable = variable

class IfC(KItem):
    def __init__(self, subsequent, alternative):
        self.subsequent = subsequent
        self.alternative = alternative

class Cfg(object):
    def __init__(self, k, state):
        self.k = k 
        self.state = state 
        self.stuck = False

def sumPgm(size):
    return Pgm(['n', 'sum'], Seq(Assign('n', ACon(size)), Seq(Assign('sum', ACon(0)), While(Not(Le(AVar('n'), ACon(0))), Seq(Assign('sum', Add(AVar('sum'), AVar('n'))), Assign('n', Add(AVar('n'), ACon(-1))))))))

# 
def step(cfg):
    k = cfg.k 
    state = cfg.state

    if len(k) == 0:
        cfg.stuck = True
        return
    
    task = k.pop()
    if isinstance(task, AVar):
        v = task.v 
        if v in state:
            k.append(ACon(state[v]))
        else:
            cfg.stuck = True
    elif isinstance(task, Div):
        x = task.x 
        y = task.y 
        if not isinstance(x, ACon):
            k.append(DivL(y))
            k.append(x)
        elif not isinstance(y, ACon):
            k.append(DivR(x))
            k.append(y)
        elif y.n == 0:
            cfg.stuck = True 
        else:
            k.append(ACon(x.n / y.n))
    elif isinstance(task, Add):
        x = task.x 
        y = task.y 
        if not isinstance(x, ACon):
            k.append(AddL(y))
            k.append(x)
        elif not isinstance(y, ACon):
            k.append(AddR(x))
            k.append(y)
        else:
            k.append(ACon(x.n + y.n))
    elif isinstance(task, Le):
        x = task.x 
        y = task.y 
        if not isinstance(x, ACon):
            k.append(LeL(y))
            k.append(x)
        elif not isinstance(y, ACon):
            k.append(LeR(x))
            k.append(y)
        else:
            k.append(BCon(x.n <= y.n))
    elif isinstance(task, Not):
        b = task.b 
        if isinstance(b, BCon):
            k.append(BCon(not b.b))
        else:
            k.append(NotF())
            k.append(b)
    elif isinstance(task, And):
        x = task.x 
        y = task.y 
        if not isinstance(x, BCon):
            k.append(AndL(y))
            k.append(x)
        else:
            if x.b:
                k.append(y)
            else:
                k.append(x)
    elif isinstance(task, Assign):
        variable = task.variable
        value = task.value 
        if isinstance(value, ACon):
            state[variable] = value.n
        else:
            k.append(AssignR(variable))
            k.append(value) 
    elif isinstance(task, Seq):
        k.append(task.s2)
        k.append(task.s1)
    elif isinstance(task, Skip):
        pass
    elif isinstance(task, If):
        condition = task.condition
        subsequent = task.subsequent 
        alternative = task.alternative 
        if isinstance(condition, BCon):
            if condition.b:
                k.append(subsequent)
            else:
                k.append(alternative)
        else:
            k.append(IfC(subsequent, alternative))
            k.append(condition)
    elif isinstance(task, While):
        condition = task.condition 
        body = task.body 
        k.append(If(condition, Seq(body, task), Skip()))
    elif isinstance(task, Pgm):
        variables = task.vars 
        body = task.body 
        if len(variables) == 0:
            k.append(body)
        else:
            variable = variables.pop()
            state[variable] = 0
            k.append(task)
    elif isinstance(task, ACon):
        n = task
        task = k.pop()
        if isinstance(task, DivL):
            k.append(Div(n, task.v))
        elif isinstance(task, DivR):
            k.append(Div(task.v, n))
        elif isinstance(task, AddL):
            k.append(Add(n, task.v))
        elif isinstance(task, AddR):
            k.append(Add(task.v, n))
        elif isinstance(task, LeL):
            k.append(Le(n, task.v))
        elif isinstance(task, LeR):
            k.append(Le(task.v, n))
        elif isinstance(task, AssignR):
            k.append(Assign(task.variable, n))
        else:
            cfg.stuck = True
    elif isinstance(task, BCon):
        b = task 
        task = k.pop()
        if isinstance(task, NotF):
            k.append(Not(b))
        elif isinstance(task, AndL):
            k.append(And(b, task.v))
        elif isinstance(task, IfC):
            k.append(If(b, task.subsequent, task.alternative))
        else:
            cfg.stuck = True
    else:
        cfg.stuck = True

        
def test(size):
    return run(sumPgm(size))

def run(p):
    cfg = Cfg([p], {})
    while True:
        step(cfg)
        if cfg.stuck:
            return cfg

def main():
    n = 10000 
    start_time = current_milli_time()
    cfg = test(n)
    print(cfg.state)
    print "Execution time in ms: "
    print (current_milli_time() - start_time)

main()

File: go_imp/imp.go
package main

import (
	"fmt"
	"time"
)

type kitem interface{}

type acon struct {
	n int64
}

type avar struct {
	v string
}

type div struct {
	x kitem
	y kitem
}

type add struct {
	x kitem
	y kitem
}

type bcon struct {
	b bool
}

type le struct {
	x kitem
	y kitem
}

type not struct {
	b kitem
}

type and struct {
	x kitem
	y kitem
}

type assign struct {
	variable string
	value    kitem
}

type _if struct {
	condition   kitem
	subsequent  kitem
	alternative kitem
}

type _while struct {
	condition kitem
	body      kitem
}

type seq struct {
	s1 kitem
	s2 kitem
}

type skip struct{}

type pgm struct {
	vars []string
	body kitem
}

type divl struct {
	v kitem
}

type divr struct {
	v kitem
}

type addl struct {
	v kitem
}

type addr struct {
	v kitem
}

type lel struct {
	v kitem
}

type ler struct {
	v kitem
}

type notf struct{}

type andl struct {
	v kitem
}

type assignr struct {
	variable string
}

type ifc struct {
	subsequent  kitem
	alternative kitem
}

type cfg struct {
	k     []kitem
	state map[string]int64
	stuck bool
}

func step(c *cfg) {
	k := c.k
	state := c.state
	if len(k) == 0 {
		c.stuck = true
		return
	}
	task := k[len(k)-1]
	k = k[:len(k)-1]

	// fmt.Println(c)

	switch task.(type) {
	case avar:
		v := task.(avar).v
		if val, ok := state[v]; ok {
			k = append(k, acon{val})
		} else {
			c.stuck = true
		}
	case div:
		x := task.(div).x
		y := task.(div).y
		if _, ok := x.(acon); !ok {
			k = append(k, divl{y}, x)
		} else if _, ok := y.(acon); !ok {
			k = append(k, divr{x}, y)
		} else if y.(acon).n == 0 {
			c.stuck = true
		} else {
			k = append(k, acon{x.(acon).n / y.(acon).n})
		}
	case add:
		x := task.(add).x
		y := task.(add).y
		if _, ok := x.(acon); !ok {
			k = append(k, addl{y}, x)
		} else if _, ok := y.(acon); !ok {
			k = append(k, addr{x}, y)
		} else {
			k = append(k, acon{x.(acon).n + y.(acon).n})
		}
	case le:
		x := task.(le).x
		y := task.(le).y
		if _, ok := x.(acon); !ok {
			k = append(k, lel{y}, x)
		} else if _, ok := y.(acon); !ok {
			k = append(k, ler{x}, y)
		} else {
			k = append(k, bcon{x.(acon).n <= y.(acon).n})
		}
	case not:
		b := task.(not).b
		if _, ok := b.(bcon); ok {
			k = append(k, bcon{!b.(bcon).b})
		} else {
			k = append(k, notf{}, b)
		}
	case and:
		x := task.(and).x
		y := task.(and).y
		if _, ok := x.(bcon); !ok {
			k = append(k, andl{y}, x)
		} else {
			if x.(bcon).b {
				k = append(k, y)
			} else {
				k = append(k, x)
			}
		}
	case assign:
		variable := task.(assign).variable
		value := task.(assign).value
		if _, ok := value.(acon); ok {
			state[variable] = value.(acon).n
		} else {
			k = append(k, assignr{variable}, value)
		}
	case seq:
		k = append(k, task.(seq).s2, task.(seq).s1)
	case skip:
	case _if:
		condition := task.(_if).condition
		subsequent := task.(_if).subsequent
		alternative := task.(_if).alternative
		if _, ok := condition.(bcon); ok {
			if condition.(bcon).b {
				k = append(k, subsequent)
			} else {
				k = append(k, alternative)
			}
		} else {
			k = append(k, ifc{subsequent, alternative}, condition)
		}
	case _while:
		condition := task.(_while).condition
		body := task.(_while).body
		k = append(k, _if{condition, seq{body, task}, skip{}})
	case pgm:
		p := task.(pgm)
		variables := p.vars
		body := p.body
		if len(variables) == 0 {
			k = append(k, body)
		} else {
			variable, variables := variables[0], variables[1:]
			state[variable] = 0
			p.vars = variables
			k = append(k, p)
		}
	case acon:
		n := task
		task, k = k[len(k)-1], k[:len(k)-1]
		switch task.(type) {
		case divl:
			k = append(k, div{n, task.(divl).v})
		case divr:
			k = append(k, div{task.(divr).v, n})
		case addl:
			k = append(k, add{n, task.(addl).v})
		case addr:
			k = append(k, add{task.(addr).v, n})
		case lel:
			k = append(k, le{n, task.(lel).v})
		case ler:
			k = append(k, le{task.(ler).v, n})
		case assignr:
			k = append(k, assign{task.(assignr).variable, n})
		default:
			c.stuck = true
		}
	case bcon:
		b := task
		task, k = k[len(k)-1], k[:len(k)-1]
		switch task.(type) {
		case notf:
			k = append(k, not{b})
		case andl:
			k = append(k, and{b, task.(andl).v})
		case ifc:
			k = append(k, _if{b, task.(ifc).subsequent, task.(ifc).alternative})
		default:
			c.stuck = true
		}
	default:
		c.stuck = true
	}

	c.k = k
	c.state = state

	// fmt.Println("enter here", k)
	// c.stuck = true
}

func sumPgm(size int64) pgm {
	var vars = []string{"n", "sum"}
	var body = seq{assign{"n", acon{size}},
		seq{assign{"sum", acon{0}},
			_while{not{le{avar{"n"}, acon{0}}},
				seq{assign{"sum", add{avar{"sum"}, avar{"n"}}},
					assign{"n", add{avar{"n"}, acon{-1}}}}}}}

	return pgm{
		vars: vars,
		body: body}
}

func run(p pgm) cfg {
	state := make(map[string]int64)
	k := make([]kitem, 0, 5)
	k = append(k, p)
	c := cfg{k: []kitem{p}, state: state, stuck: false}

	for {
		step(&c)
		if c.stuck {
			return c
		}
	}
}

func test(size int64) {
	start := time.Now()
	c := run(sumPgm(size))
	elapsed := time.Since(start)
	fmt.Println(c)
	fmt.Println("Execution time: ", elapsed.Seconds()*1000, " ms")
}

func main() {
	test(10000000)
}
